forem/forem
# Refactor edge caching logic to live in EdgeCache::Service

## What type of PR is this? (check all applicable)


## Description

When we first added support for multiple caching providers in https://github.com/forem/forem/pull/9857, I added all the logic into the `CacheBuster.bust` method. This was fine initially as we ensured that we weren't breaking anything, but that module has a lot of different kinds of logic living in there now when it really probably doesn't need to.

This PR adds a new `EdgeCache::Service` that is responsible for knowing which edge caching provider we are using (Fastly or Nginx), and contains the logic for pinging both.

This allows the `CacheBuster` module to be much cleaner, containing only the logic to decide which path(s) to bust. The `EdgeCache::Service`, on the other hand, encapsulates the logic around knowing which provider we should be using to purge a cache, and whether or not that provider is configured + available.

```bash
CacheBuster.bust(some-path-here)
=> #<EdgeCache::Service:0x00007fd6ce30aff0 @path="some-path-here", @provider="your-configured-provider-here">
```

## Related Tickets & Documents

Improving upon https://github.com/forem/forem/pull/9857.

## QA Instructions, Screenshots, Recordings

⚠️  **To QA this PR, you will need have Openresty set up locally on your machine, which is powered by Docker. Please see the [LFSS repo](https://github.com/forem/lfss) for details on how to get yourself situated and set up.**


🚨 **To QA this PR, you will also need two ENV variables set in your local `.env`: `OPENRESTY_PROTOCOL="http://"` and `OPENRESTY_DOMAIN="localhost:9090"`. Be sure to confirm that these are accessible to you from `ApplicationConfig`, otherwise you won't be able to hit Nginx locally!!**

To test this out, you'll want 4 windows open in your terminal:
A. LFSS window with `docker-compose up` running
B. A window to make `curl` requests (to `HIT` the cached content)
C. A window with your `rails server` running
D. A window for your `rails console`

In window **B**, make a `curl` request to an article in your local environment like so:
```bash
➜  lfss git:(edge) curl -I http://localhost:9090/vaidehijoshi/vaidehi-s-test-post-1878
HTTP/1.1 200 OK
Server: openresty
Date: Thu, 17 Sep 2020 18:50:49 GMT
Content-Type: text/html; charset=utf-8
Connection: keep-alive
X-Frame-Options: SAMEORIGIN
X-XSS-Protection: 1; mode=block
X-Content-Type-Options: nosniff
X-Download-Options: noopen
X-Permitted-Cross-Domain-Policies: none
Referrer-Policy: strict-origin-when-cross-origin
Cache-Control: public, no-cache
Surrogate-Control: max-age=86400, stale-if-error=26400
Surrogate-Key: articles/27
Vary: Accept-Encoding
ETag: W/"722b30c5411854b65c1f645c813b599d"
Content-Security-Policy: connect-src 'self' https: http://localhost:3035 ws://localhost:3035
X-Request-Id: 12d1ae7d-3ee5-4887-8661-659a927302ab
X-Runtime: 1.110368
X-Cache-Status: MISS
```
You should see a cache `MISS` from that request.

If you look at window **A**, you should also see a cache `MISS`:
```bash
➜  lfss git:(edge) docker-compose up
Starting openresty ... done

openresty    | 172.18.0.1 - - [17/Sep/2020:19:55:46 +0000] "HEAD /vaidehijoshi/vaidehi-s-test-post-1878 HTTP/1.1" 200 0 "-" "curl/7.54.0" "-" "MISS"
```

But, if you make that _same_ `curl` request again, you'll see that your `curl` request now responds with `X-Cache-Status: HIT`, and if you look at window **A** again, you'll see that you now are hitting the cached version of the content:
```bash
➜  lfss git:(edge) docker-compose up
Starting openresty ... done

openresty    | 172.18.0.1 - - [17/Sep/2020:19:55:46 +0000] "HEAD /vaidehijoshi/vaidehi-s-test-post-1878 HTTP/1.1" 200 0 "-" "curl/7.54.0" "-" "MISS"
openresty    | 172.18.0.1 - - [17/Sep/2020:19:56:18 +0000] "HEAD /vaidehijoshi/vaidehi-s-test-post-1878 HTTP/1.1" 200 0 "-" "curl/7.54.0" "-" "HIT"
```

Now, open up window **D**, your `rails console`, and directly invoke the `CacheBuster.bust`, passing in the _path to the article_ you have been `curl`ing to as the argument:
```bash
➜  forem git:(vaidehijoshi/create-edgecache-service) rails c
[1] pry(main)> CacheBuster.bust("/vaidehijoshi/vaidehi-s-test-post-1878")
=> #<EdgeCache::Service:0x00007fd6ce30aff0 @path="/vaidehijoshi/vaidehi-s-test-post-1878", @provider="nginx">
```

**Notice that, because you have the OPENRESTY env vars configured, you can see that the `EdgeCache::Service`, the return value of `CacheBuster.bust` has its `provider` listed as `"nginx"`, and it's `path` should be the article path that you passed to it.**


The request should complete successfully, and you should see the `PURGE` request come through in window **A**:
```bash
openresty    | 172.18.0.1 - - [17/Sep/2020:19:55:46 +0000] "HEAD /vaidehijoshi/vaidehi-s-test-post-1878 HTTP/1.1" 200 0 "-" "curl/7.54.0" "-" "MISS"
openresty    | 172.18.0.1 - - [17/Sep/2020:19:56:18 +0000] "HEAD /vaidehijoshi/vaidehi-s-test-post-1878 HTTP/1.1" 200 0 "-" "curl/7.54.0" "-" "HIT"
openresty    | 172.18.0.1 - - [17/Sep/2020:19:57:39 +0000] "GET / HTTP/1.1" 200 71174 "-" "Ruby" "-" "HIT"
openresty    | 172.18.0.1 - - [17/Sep/2020:19:57:39 +0000] "PURGE /vaidehijoshi/vaidehi-s-test-post-1878 HTTP/1.1" 200 13 "-" "Ruby" "-" "-"
```

Now, you can confirm that caching still works by `curl`ing to that same article again. You should see a `MISS` on the first `curl`, and a `HIT` on the second one:
```bash
openresty    | 172.18.0.1 - - [17/Sep/2020:19:57:39 +0000] "PURGE /vaidehijoshi/vaidehi-s-test-post-1878 HTTP/1.1" 200 13 "-" "Ruby" "-" "-"
openresty    | 172.18.0.1 - - [17/Sep/2020:19:59:19 +0000] "HEAD /vaidehijoshi/vaidehi-s-test-post-1878 HTTP/1.1" 200 0 "-" "curl/7.54.0" "-" "MISS"
openresty    | 172.18.0.1 - - [17/Sep/2020:19:59:20 +0000] "HEAD /vaidehijoshi/vaidehi-s-test-post-1878 HTTP/1.1" 200 0 "-" "curl/7.54.0" "-" "HIT"
```

Success!! 🎉 


You should also test that the `EdgeCache::Service` doesn't try to ping a caching provider unless it has been configured. To do this, you should _remove_ any `OPENRESTY_` env vars entirely, and restart your rails app/rails console. Then, you can test that busting that same article path results in no change:
```bash
➜  forem git:(vaidehijoshi/create-edgecache-service) ✗ rails c
Loading development environment (Rails 6.0.3.3)
[1] pry(main)> CacheBuster.bust("/vaidehijoshi/vaidehi-s-test-post-1878")
Unset ENV variable: OPENRESTY_PROTOCOL.
=> nil
```

Notice that you do not get an `EdgeCache::Service` instance returned to you if there is no caching provider set!

## Added tests?


## Added to documentation?

- [x] ~no documentation needed~ inline documentation

## Are there any post deployment tasks we need to perform?

Make sure that Fastly behaves correctly with regards to purging caches on production!

## What gif best describes this PR or how it makes you feel?

![veep gif](https://media3.giphy.com/media/11FtBDzEDrUUHC/giphy.webp?cid=5a38a5a2my6tlcyob5mmy5bxrvzanyj3lnyast12fco13jyp&rid=giphy.webp)
